harness/determined/common/experimental/checkpoint/_checkpoint.py

```diff
@@ -147,31 +153,48 @@ def download(self, path: Optional[str] = None) -> str:
                 raise NotImplementedError("Non-training checkpoints cannot be downloaded")
 
             checkpoint_storage = self.training.experiment_config["checkpoint_storage"]
-            if checkpoint_storage["type"] == "shared_fs":
-                src_ckpt_dir = self._find_shared_fs_path(checkpoint_storage)
-                shutil.copytree(str(src_ckpt_dir), str(local_ckpt_dir))
-            else:
-                local_ckpt_dir.mkdir(parents=True, exist_ok=True)
-                manager = storage.build(
-                    checkpoint_storage,
-                    container_path=None,
-                )
-                if not isinstance(
-                    manager,
-                    (
-                        storage.S3StorageManager,
-                        storage.GCSStorageManager,
-                        storage.AzureStorageManager,
-                    ),
-                ):
-                    raise AssertionError(
-                        "Downloading from Azure, S3 or GCS requires the experiment to be "
-                        "configured with Azure, S3 or GCS checkpointing, {} found instead".format(
-                            checkpoint_storage["type"]
+            try:
```

Comment:
I _think_ what you have here is correct, but I think the structure of if statements and try statements is not making the behavior of the code very clear.

How about something like this pseudocode:
```python
if mode == AUTO:
     try:
         self._download_direct()
     except NoCloudAccess as e1:
         logging.info("unable to download directly, proxying download through master")
         self.download_via_master
elif mode == VIA_MASTER:
     self._download_via_master()
elif mode == DIRECT:
     self._download_direct()
else:
    raise ValueError(mode)
```

A structure like this is clearer than the one-big-try-except pattern because it's not clear from the one-big-try-except which codepaths might result in the specified exception.  Also it is very clear that the mode provided by the user is explicitly controlling the codepath (in a way that a bool isn't really capable of)